File: app/src/main/java/com/kirianov/multisim/Slot.java
package com.kirianov.multisim;

import java.util.List;

public class Slot {

    private String imei;
    private String imsi;
    // private String msisdn; // not posible now

    private int simState;
    private String simSerialNumber;
    private String simOperator;
    private String simOperatorName;
    private String simCountryIso;

    private String networkOperator;
    private String networkOperatorName;
    private String networkCountryIso;
    private int networkType;
    // careful find networkTypeName because it will be different with networkType on same devices
    private boolean networkRoaming;


    public Slot() {
        simState = -1;
    }

    public String getImei() {
        return imei;
    }

    public int getSimState() {
        return simState;
    }

    public String getImsi() {
        return imsi;
    }

    public String getSimSerialNumber() {
        return simSerialNumber;
    }

    public String getSimOperator() {
        return simOperator;
    }

    public String getSimOperatorName() {
        return simOperatorName;
    }

    public String getSimCountryIso() {
        return simCountryIso;
    }

    public String getNetworkOperator() {
        return networkOperator;
    }

    public String getNetworkOperatorName() {
        return networkOperatorName;
    }

    public String getNetworkCountryIso() {
        return networkCountryIso;
    }

    public int getNetworkType() {
        return networkType;
    }

    public boolean isNetworkRoaming() {
        return networkRoaming;
    }


    void setImei(String imei) {
        this.imei = imei;
    }

    void setSimState(Integer simState) {
        if (simState == null) {
            this.simState = -1;
            return;
        }
        this.simState = simState;
    }

    void setImsi(String imsi) {
        this.imsi = imsi;
    }

    void setSimSerialNumber(String simSerialNumber) {
        this.simSerialNumber = simSerialNumber;
    }

    void setSimOperator(String simOperator) {
        this.simOperator = simOperator;
    }

    void setSimOperatorName(String simOperatorName) {
        this.simOperatorName = simOperatorName;
    }

    void setSimCountryIso(String simCountryIso) {
        this.simCountryIso = simCountryIso;
    }

    void setNetworkOperator(String networkOperator) {
        this.networkOperator = networkOperator;
    }

    void setNetworkOperatorName(String networkOperatorName) {
        this.networkOperatorName = networkOperatorName;
    }

    void setNetworkCountryIso(String networkCountryIso) {
        this.networkCountryIso = networkCountryIso;
    }

    void setNetworkType(Integer networkType) {
        if (networkType == null) {
            this.networkType = 0;
            return;
        }
        this.networkType = networkType;
    }

    void setNetworkRoaming(Boolean networkRoaming) {
        if (networkRoaming == null) {
            this.networkRoaming = false;
            return;
        }
        this.networkRoaming = networkRoaming;
    }

    public String toString() {
        return "imei=[" + imei +
                "] imsi=[" + imsi +
                "] simState=[" + simState +
                "] simSerialNumber=[" + simSerialNumber +
                "] simOperator=[" + simOperator +
                "] simOperatorName=[" + simOperatorName +
                "] simCountryIso=[" + simCountryIso +
                "] networkOperator=[" + networkOperator +
                "] networkOperatorName=[" + networkOperatorName +
                "] networkCountryIso=[" + networkCountryIso +
                "] networkType=[" + networkType +
                "] networkRoaming=[" + networkRoaming + "]";
    }

    public boolean compare(Slot slot) {
        if (slot == null) return false;
        // if (("" + getImei()).compareTo("" + slot.getImei()) == 0) return true;
        return ((("" + getImei()).compareTo("" + slot.getImei()) == 0) &&
                (("" + getImsi()).compareTo("" + slot.getImsi()) == 0) &&
                (("" + getSimSerialNumber()).compareTo("" + slot.getSimSerialNumber()) == 0) // &&
                // (("" + getSimOperator()).compareTo("" + slot.getSimOperator()) == 0) &&
                // (("" + getSimOperatorName()).compareTo("" + slot.getSimOperatorName()) == 0) &&
                // (("" + getSimCountryIso()).compareTo("" + slot.getSimCountryIso()) == 0) &&
                // (getSimState() == slot.getSimState()) &&
                // (("" + getSimSerialNumber()).compareTo("" + slot.getSimSerialNumber()) == 0) &&
                // (("" + getNetworkOperator()).compareTo("" + slot.getNetworkOperator()) == 0) &&
                // (("" + getNetworkOperatorName()).compareTo("" + slot.getNetworkOperatorName()) == 0) &&
                // (("" + getNetworkCountryIso()).compareTo("" + slot.getNetworkCountryIso()) == 0) &&
                // (getNetworkType() == slot.getNetworkType()) &&
                // (isNetworkRoaming() == slot.isNetworkRoaming())
        );
    }

    public int indexIn(List<Slot> slots) {
        if (slots == null) return -1;
        for (int i = 0; i < slots.size(); i++)
            if (this.compare(slots.get(i)))
                return i;
        return -1;
    }

    public boolean containsIn(List<Slot> slots) {
        if (slots == null) return false;
        for (Slot slot : slots)
            if (this.compare(slot))
                return true;
        return false;
    }

}

File: app/src/main/java/com/kirianov/multisim/Slot.kt
package com.kirianov.multisim

class Slot {

    var imei: String? = null

    var imsi: String? = null

    var simState = -1

    var simSerialNumber: String? = null

    var simOperator: String? = null

    var simOperatorName: String? = null

    var simCountryIso: String? = null

    var networkOperator: String? = null

    var networkOperatorName: String? = null

    var networkCountryIso: String? = null

    var networkType = 0

    // careful find networkTypeName because it will be different with networkType on same devices
    var isNetworkRoaming = false

    // private String msisdn; // not posible now

    fun setSimState(state: Int?) {
        if (state == null) {
            simState = -1
            return
        }
        simState = state
    }

    fun setNetworkType(type: Int?) {
        if (type == null) {
            networkType = 0
            return
        }
        networkType = type
    }

    fun setNetworkRoaming(networkRoaming: Boolean?) {
        if (networkRoaming == null) {
            isNetworkRoaming = false
            return
        }
        isNetworkRoaming = networkRoaming
    }

    override fun toString(): String {
        return "imei=[" + imei +
            "] imsi=[" + imsi +
            "] simState=[" + simState +
            "] simSerialNumber=[" + simSerialNumber +
            "] simOperator=[" + simOperator +
            "] simOperatorName=[" + simOperatorName +
            "] simCountryIso=[" + simCountryIso +
            "] networkOperator=[" + networkOperator +
            "] networkOperatorName=[" + networkOperatorName +
            "] networkCountryIso=[" + networkCountryIso +
            "] networkType=[" + networkType +
            "] networkRoaming=[" + isNetworkRoaming + "]"
    }

    fun compare(slot: Slot?): Boolean {
        return if (slot == null) {
            false
        } else {
            imei == slot.imei && imsi == slot.imsi && simSerialNumber == slot.simSerialNumber
        }// &&
        // (("" + getSimOperator()).compareTo("" + slot.getSimOperator()) == 0) &&
        // (("" + getSimOperatorName()).compareTo("" + slot.getSimOperatorName()) == 0) &&
        // (("" + getSimCountryIso()).compareTo("" + slot.getSimCountryIso()) == 0) &&
        // (getSimState() == slot.getSimState()) &&
        // (("" + getSimSerialNumber()).compareTo("" + slot.getSimSerialNumber()) == 0) &&
        // (("" + getNetworkOperator()).compareTo("" + slot.getNetworkOperator()) == 0) &&
        // (("" + getNetworkOperatorName()).compareTo("" + slot.getNetworkOperatorName()) == 0) &&
        // (("" + getNetworkCountryIso()).compareTo("" + slot.getNetworkCountryIso()) == 0) &&
        // (getNetworkType() == slot.getNetworkType()) &&
        // (isNetworkRoaming() == slot.isNetworkRoaming())
    }

    fun indexIn(slots: List<Slot>?): Int {
        if (slots == null) {
            return -1
        }
        for (i in slots.indices) {
            if (compare(slots[i])) {
                return i
            }
        }
        return -1
    }

    fun containsIn(slots: List<Slot>?): Boolean {
        if (slots == null) {
            return false
        }
        for (slot in slots) {
            if (compare(slot)) {
                return true
            }
        }
        return false
    }
}
